### PR TITLE
chore: Uses `use_replication_spec_per_shard` in data sources

### PR DESCRIPTION
**Pull request by lantoli**
## Description

Uses `use_replication_spec_per_shard` in data sources

Link to any related issue(s): CLOUDP-290307

## Type of change:

- [ ] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR. A migration guide must be created or updated if the new feature will go in a major version.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR. A migration guide must be created or updated.
- [ ] This change requires a documentation update
- [ ] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contributing guides](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/contributing/README.md)
- [x] I have checked that this change does not generate any credentials and that **they are NOT accidentally logged anywhere**.
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code
- [x] If changes include deprecations or removals I have added appropriate changelog entries.
- [x] If changes include removal or addition of 3rd party GitHub actions, I updated our internal document. Reach out to the APIx Integration slack channel to get access to the internal document.

### CODE REVIEW

File: internal/service/advancedclustertpf/resource_compatiblity.go [hunk 1]
```diff
@@ -53,12 +63,23 @@ func resolveAPIInfo(ctx context.Context, plan *TFModel, diags *diag.Diagnostics,
 		diags.AddError("resolveContainerIDs failed", err.Error())
 		return nil
 	}
+	var (
+		legacySchema      bool
+		zoneNameNumShards map[string]int64
+	)
+	if overrideUsingLegacySchema {
+		legacySchema = true
+		zoneNameNumShards = numShardsMapFromOldAPI(clusterRespOld)
```
Comment (lantoli) [inline internal/service/advancedclustertpf/resource_compatiblity.go]: plan is empty in Read when overriding legacy, so we get it from the old API

File: internal/service/advancedclustertpf/resource_compatiblity.go [hunk 2]
```diff
@@ -37,13 +39,21 @@ func findNumShardsUpdates(ctx context.Context, state, plan *TFModel, diags *diag
 	return planCounts
 }
 
-func resolveAPIInfo(ctx context.Context, plan *TFModel, diags *diag.Diagnostics, clusterLatest *admin.ClusterDescription20240805, client *config.MongoDBClient) *ExtraAPIInfo {
+func resolveAPIInfo(ctx context.Context, diags *diag.Diagnostics, client *config.MongoDBClient, plan *TFModel, clusterLatest *admin.ClusterDescription20240805, overrideUsingLegacySchema bool) *ExtraAPIInfo {
```
Comment (EspenAlbert) [inline internal/service/advancedclustertpf/resource_compatiblity.go]: [nit] `forceLegacySchema` less verbose?

File: internal/service/advancedclustertpf/resource_compatiblity.go [hunk 3]
```diff
@@ -37,13 +39,21 @@ func findNumShardsUpdates(ctx context.Context, state, plan *TFModel, diags *diag
...
 	rootDiskSize := conversion.NilForUnknown(plan.DiskSizeGB, plan.DiskSizeGB.ValueFloat64Pointer())
 	projectID := plan.ProjectID.ValueString()
-	zoneNameSpecIDs, asymmetricShardUnsupported, err := getReplicationSpecIDsFromOldAPI(ctx, projectID, plan.Name.ValueString(), client.AtlasV220240530.ClustersApi)
+	clusterName := plan.Name.ValueString()
+	asymmetricShardUnsupported := false
+
+	api20240530 := client.AtlasV220240530.ClustersApi
+	clusterRespOld, _, err := api20240530.GetCluster(ctx, projectID, clusterName).Execute()
```
Comment (EspenAlbert) [inline internal/service/advancedclustertpf/resource_compatiblity.go:50]: Nice refactor!

**Review by EspenAlbert — APPROVED** [review]
Nice. If we merge this now, we don't have to undo changes later here in this [PR](https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2895), we'll have to merge master to that branch

File: internal/service/advancedclustertpf/resource_compatiblity.go (hunk 1 above)
Comment (EspenAlbert) [reply internal/service/advancedclustertpf/resource_compatiblity.go]: Add this comment to the code?

Comment (lantoli) [reply internal/service/advancedclustertpf/resource_compatiblity.go]: added in: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2896/commits/d39cf2f073eec031e73dc8f9b124673b5aca6855

File: internal/service/advancedclustertpf/resource_compatiblity.go (hunk 2 above)
Comment (lantoli) [reply internal/service/advancedclustertpf/resource_compatiblity.go]: changed in: https://github.com/mongodb/terraform-provider-mongodbatlas/pull/2896/commits/d39cf2f073eec031e73dc8f9b124673b5aca6855

File: internal/service/advancedclustertpf/resource_compatiblity.go [hunk 4]
```diff
@@ -53,14 +64,20 @@ func resolveAPIInfo(ctx context.Context, plan *TFModel, diags *diag.Diagnostics,
...
-		UsingLegacySchema:          usingLegacySchema(ctx, plan.ReplicationSpecs, diags),
-		ZoneNameNumShards:          numShardsMap(ctx, plan.ReplicationSpecs, diags),
 		RootDiskSize:               rootDiskSize,
-		ZoneNameReplicationSpecIDs: zoneNameSpecIDs,
+		ZoneNameReplicationSpecIDs: replicationSpecIDsFromOldAPI(clusterRespOld),
 		AsymmetricShardUnsupported: asymmetricShardUnsupported,
 	}
+	if forceLegacySchema {
+		info.UsingLegacySchema = true
+		info.ZoneNameNumShards = numShardsMapFromOldAPI(clusterRespOld) // plan is empty in data source Read when forcing legacy, so we get num_shards from the old API
```
Comment (maastha) [inline internal/service/advancedclustertpf/resource_compatiblity.go:75]: [q] can you elaborate on what you mean by "plan is empty" here? Simply that num_shards won't have a value or something else?

Comment (lantoli) [reply internal/service/advancedclustertpf/resource_compatiblity.go:75]: in a data source the plan is empty, it only has the attributes provided by the user in the TF files. for example replication_specs will be empty

**Pull request merged**